### PR TITLE
Add client to fetch Application/Schedule/Preferences details 

### PR DESCRIPTION
**Pull request by wyzhang**
Introduce Abstract*Fetcher interface and Remote*Fetcher implementation that fetch Application, Schedule and Preferences detials. 

Next change will be switching ProfileMetadataMessageProcessor to use these clients instead of currently accessing levelDB directly

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.id.ApplicationId;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Interface for fetching {@code ApplicationDetail}
+ */
+public interface AbstractApplicationDetailFetcher {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: I think we can remove the 'Abstract' and just have ApplicationDetailFetcher.

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: is the plan to have ApplicationLifecycleService implement this interface?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Get the application detail for the given application id
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: javadoc should include information about when the different exceptions are thrown

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+   */
+  ApplicationDetail get(ApplicationId appId) throws IOException, NotFoundException;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: should this throw ApplicationNotFoundException instead of NotFoundException? Or will it be throwing other types of NotFoundException as well?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Get details of all applications in the given namespace
+   */
+  List<ApplicationDetail> list(String namespace) throws IOException, NotFoundException;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: similar comments here. When does it throw NotFoundException? when the namespace doesn't exist? Then shouldn't it throw NamespaceNotFoundException?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  public ApplicationDetail get(ApplicationId appId) throws IOException, NotFoundException {
+    String url = String.format("namespaces/%s/apps/%s/versions/%s",
+                               appId.getNamespace(), appId.getApplication(), appId.getVersion());
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    HttpResponse httpResponse;
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(appId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: the request could have failed with a 404 if the namespace doesn't exist as well so this isn't always going to be accurate. Unfortunately I don't think there is a way to differentiate the two without larger changes, so I think it's better to just throw a NotFoundException where the exception message is the response body.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    ObjectResponse<List<ApplicationDetail>> objectResponse = ObjectResponse.fromJsonBody(
+      httpResponse, new TypeToken<List<ApplicationDetail>>() {
+      }.getType(), GSON);
+    return objectResponse.getResponseObject();
+  }
+
+  private HttpResponse execute(HttpRequest request) throws IOException, NotFoundException {
+    HttpResponse httpResponse = remoteClient.execute(request);
+    if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new NotFoundException("Not found");
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: should use the response body instead of 'Not found', as it should be more descriptive

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    String url = String.format("namespaces/%s/apps", namespace);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    HttpResponse httpResponse;
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(new NamespaceId(namespace));
+    }
+    ObjectResponse<List<ApplicationDetail>> objectResponse = ObjectResponse.fromJsonBody(
+      httpResponse, new TypeToken<List<ApplicationDetail>>() {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: this Type can be a private static final constant instead of created in each call

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+   * Get details of all applications in the given namespace
+   */
+  public List<ApplicationDetail> list(String namespace) throws IOException, NotFoundException {
+    String url = String.format("namespaces/%s/apps", namespace);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    HttpResponse httpResponse;
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(new NamespaceId(namespace));
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: should throw a NamespaceNotFoundException

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    return objectResponse.getResponseObject();
+  }
+
+  private HttpResponse execute(HttpRequest request) throws IOException, NotFoundException {
+    HttpResponse httpResponse = remoteClient.execute(request);
+    if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new NotFoundException("Not found");
+    }
+    if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+      throw new IOException(String.format("Request failed %s", httpResponse.getResponseBodyAsString()));
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: just use the response body by itself without the 'Request failed', otherwise the message can end up looking weird.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * Fetch preferences via REST API calls (using internal endpoint {@code INTERNAL_API_VERSION_3})
+ */
+public class RemotePreferencesFetcherInternal implements AbstractPreferencesFetcher {
+  private static final Gson GSON = new Gson();
+
+  private final RemoteClient remoteClient;
+
+  @Inject
+  public RemotePreferencesFetcherInternal(final DiscoveryServiceClient discoveryClient) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: doesn't need to be final

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * Fetch application detail via REST API calls
+ */
+public class RemoteApplicationDetailFetcher implements AbstractApplicationDetailFetcher {
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder()).create();
+
+  private final RemoteClient remoteClient;
+
+  @Inject
+  public RemoteApplicationDetailFetcher(final DiscoveryServiceClient discoveryClient) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: doesn't need to be final

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+  public PreferencesDetail get(EntityId entityId, boolean resolved) throws IOException, NotFoundException {
+    HttpResponse httpResponse;
+    String url = getPreferencesURI(entityId, resolved);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(entityId);
+    }
+    return GSON.fromJson(httpResponse.getResponseBodyAsString(), new TypeToken<PreferencesDetail>() {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: type can be a private static final constant

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+   * Get preferences for the given identify
+   */
+  public PreferencesDetail get(EntityId entityId, boolean resolved) throws IOException, NotFoundException {
+    HttpResponse httpResponse;
+    String url = getPreferencesURI(entityId, resolved);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(entityId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: similar comment, this is misleading because it's not always the entity that doesn't exist

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    return uri;
+  }
+
+  private HttpResponse execute(HttpRequest request) throws IOException, NotFoundException {
+    HttpResponse httpResponse = remoteClient.execute(request);
+    if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new NotFoundException("Not found");
+    }
+    if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+      throw new IOException(String.format("Request failed %s", httpResponse.getResponseBodyAsString()));
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: there is code duplication in these classes, it would be better to consolidate the common code, either in a common abstract class or by enhancing RemoteClient in some way.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternalTest.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    Map<String, String> resolvedProperites = new HashMap<>();
+    resolvedProperites.putAll(appProperties);
+    resolvedProperites.putAll(instanceProperties);
+    Assert.assertEquals(resolvedProperites, preferences.getProperties());
+    Assert.assertTrue(preferences.getResolved());
+    Assert.assertTrue(preferences.getSeqId() > 0);
+
+    // Delete the app
+    Assert.assertEquals(
+        200,
```
Comment (albertshau) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternalTest.java]: indentation

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java (hunk 1 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: Done. Removed "Abstract" 

Yes. But I am a bit leaning toward creating a LocalxxxFetcher that contains an instance of xxxService instead of have xxxService implement the interface, so it looks a bit more symmetric. Coming from C++ background, I find a class implementing multiple different interfaces a bit intuitive.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java (hunk 2 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java (hunk 4 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: Done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java (hunk 3 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: Done.

Just ApplicationNotFound. So catch NotFound and throw ApplicationNotFound.

Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java]: Changed to throw NotFound as it could be AppNotFound or NamespaceNotFound

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 5 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: Done. Changed to just NotFound

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 11 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java (hunk 12 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: Can you clarify which line of code are you referring to?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java (hunk 13 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: Fixed. Just throw NotFoundException

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java (hunk 14 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: Agreed. Added a TODO to refactor in a follow-up PR

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternalTest.java (hunk 15 above)
Comment (wyzhang) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternalTest.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 7 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: Fixed.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 9 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: Fixed

**Comment by wyzhang** [comment]
https://github.com/cdapio/cdap/pull/11903/commits/dd26dac0bfb04eb8a337f02854aa033dc479471d

Discovered that /apps and /apps/<app-id> endpoints return inconsistent info. One is a list of ApplicationRecord and the other is ApplicationDetail 

Cannot fix this inconsistency as they are public. 

Created/added new endpoints /appdetail and /appdetail/<appid>/versions/<versionid> that return ApplicatoinDetails. Updated the client to switch to use these new endpoint

**Comment by wyzhang** [comment]
build and test passed
https://builds.cask.co/browse/CDAP-DUT7117-4

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java (hunk 12 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: this line, the `new TypeToken`

Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: Oh.. Already fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 6 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: Done. Already fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 8 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: Fixed. Make list() to catch NotFound and then throw NamespaceNotFound. 

For get(appId), just throw not found as currently there is no way of telling AppNotFound vs NamespaceNotFound.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java (hunk 10 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java]: Removed.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 16]
```diff
@@ -416,7 +437,46 @@ public void getApplicationDetails(FullHttpRequest request, HttpResponder respond
...
+   *
+   * @param request     {@link HttpRequest}
+   * @param responder   {@link HttpResponse}
+   * @param namespace   the namespace to get all application details   *
+   * @param application the id of the application to get its {@link ApplicationDetail}
+   * @throws Exception if either namespace or application doesn't exist, or failed to get {@link ApplicationDetail}
+   */
+  @GET
+  @Path("/appdetail/{app-id}")
+  public void getAppDetail(HttpRequest request, HttpResponder responder,
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: why do we need this endpoint? Isn't it the same as apps/app-d?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 17]
```diff
@@ -416,7 +437,46 @@ public void getApplicationDetails(FullHttpRequest request, HttpResponder respond
...
+   * Get {@link ApplicationDetail} for a given application
+   *
+   * @param request     {@link HttpRequest}
+   * @param responder   {@link HttpResponse}
+   * @param namespace   the namespace to get all application details   *
+   * @param application the id of the application to get its {@link ApplicationDetail}
+   * @throws Exception if either namespace or application doesn't exist, or failed to get {@link ApplicationDetail}
+   */
+  @GET
+  @Path("/appdetail/{app-id}/versions/{version-id}")
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: similar question, why is this needed?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 18]
```diff
@@ -215,15 +216,21 @@ public BodyConsumer createAppVersion(HttpRequest request, HttpResponder responde
   }
 
   /**
-   * Returns a list of applications associated with a namespace.
+   * Returns a list of {@link ApplicationRecord} for all applications in the given namespace
+   *
+   * @param request         {@link HttpRequest}
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: style: our javadocs usually aren't aligned and just use a single space between param and description

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 19]
```diff
@@ -416,7 +437,46 @@ public void getApplicationDetails(FullHttpRequest request, HttpResponder respond
   }
 
   /**
-   * Decodes request coming from the {@link #getApplicationDetails(FullHttpRequest, HttpResponder, String)} call.
+   * Get {@link ApplicationDetail} for a given application
+   *
+   * @param request     {@link HttpRequest}
+   * @param responder   {@link HttpResponse}
+   * @param namespace   the namespace to get all application details   *
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: extra '*' at the end of the line. Same comment in other places in this file

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 20]
```diff
@@ -378,6 +382,23 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
...
+  /**
+   * Get a list of {@link ApplicationDetail} for all applications in the given namespace
+   *
+   * @param request   {@link HttpRequest}
+   * @param responder {@link HttpResponse}
+   * @param namespace the namespace to get all application details
+   * @throws Exception if namespace doesn't exists or failed to get all application details
+   */
+  @GET
+  @Path("/appdetail")
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: what's the reason the other app list endpoint was changed? Are we causing issues by adding back the old functionality?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 16 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Yes. It is the same as /app/<appid>. I hope at some point we can change /app/<appid> to be consistent with /apps to return ApplicationRecord instead of ApplicationDetail. 

Introduce this /appdetail/<appid> to achieve the parity, as I dislike the fact that client uses /app/<appid> in get-one case and /appdetail for get-all case.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 17 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: same as above.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 18 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: hm..the cask setting I imported set para aligned in the javadoc. 

Disabled and reformatted the code.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 19 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Removed extra "*".

Fixing the param doc alignment

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 20 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: The inconsistency between /apps returning in ApplicationDetails format while /app/<appid> etc returning ApplicationRecord has been there for quite a while (since the acquisition from what I can tell from the change history). 

Unable to see change history past the acquisition (since it was in a different repo?)

Anyway, regardless of why the inconsistency was introduced, this inconsistency looks ugly and error prone. I didn't notice this and my unit test didn't catch this due to json deserialization just ignore unknown fields, so unit tests on the client side needs to check as many fields as possible to ensure this gets caught. 

I would prefer not to use these endpoints going forward in hope of deprecating them somehow (probably not possible), but at least not adding new code use them.

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: I misunderstood the commit comments, so there isn't any change. It's not an uncommon pattern for a list to return less information than a get on the specific element returns.

Should this be an internal handler then instead of mixed in with the user facing APIs?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 16 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: we should not have duplicate endpoints in this class.

I think it's ok if there is an internal handler that has the equivalent of this endpoint, but it shouldn't be mixed in with the public APIs.

Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Done. Moved to internal endpoint

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 20 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Ack. 

Yes. Moved these to internal. Keep public unchanged.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+  /**
+   * Get a list of {@link ApplicationDetail} for all applications in the given namespace
+   *
+   * @param request   {@link HttpRequest}
+   * @param responder {@link HttpResponse}
+   * @param namespace the namespace to get all application details
+   * @throws Exception if namespace doesn't exists or failed to get all application details
+   */
+  @GET
+  @Path("/appdetail")
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java]: nit: collections should be plural -> 'appdetails'. Though I think it's better to just have it as 'apps'

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 7 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: this is not fixed yet

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 9 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: This isn't fixed either, is there a missing commit?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcherTest.java [hunk 22]
```diff
@@ -0,0 +1,143 @@
+/*
+ru * Copyright © 2020 Cask Data, Inc.
```
Comment (albertshau) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcherTest.java]: remove 'ru'

Comment (wyzhang) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcherTest.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java (hunk 21 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java]: Fixed. Changed to /apps and /app/<appid>

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 9 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: I think I might have missed committing that change. 

Fixed. Please double check.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java (hunk 7 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java]: Done. Fixed. Might have missed committing a change.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+  /**
+   * Runtime program service for running and managing programs.
+   */
+  private final ProgramRuntimeService runtimeService;
+
+  private final CConfiguration configuration;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final NamespacePathLocator namespacePathLocator;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final File tmpDir;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java:62]: Seems like an unused field? Try to use the IDE warning to get rid of unused fields.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3 + "/namespaces/{namespace-id}")
+public class AppLifecycleHttpHandlerInternal extends AbstractAppFabricHttpHandler {
+  private static final Gson GSON = new Gson();
+
+  /**
+   * Runtime program service for running and managing programs.
+   */
+  private final ProgramRuntimeService runtimeService;
+
+  private final CConfiguration configuration;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java:58]: Unused field?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/ScheduleFetcher.java [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+  /**
+   * Get schedule details for the given {@code scheduleId}
+   * @param scheduleId the id of the schedule to fetch {@code ScheduleDetail} for
+   * @return the schedule detail
+   * @throws IOException if failed to get the detail of the given schedule
+   * @throws ScheduleNotFoundException if the given schedule doesn't exist.
+   */
+  ScheduleDetail get(ScheduleId scheduleId) throws IOException, ScheduleNotFoundException;
+
+  /**
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/ScheduleFetcher.java:41]: Remove empty comment block.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java [hunk 26]
```diff
@@ -106,7 +106,7 @@ public void getAllAppDetails(HttpRequest request, HttpResponder responder,
    * @throws Exception if either namespace or application doesn't exist, or failed to get {@link ApplicationDetail}
    */
   @GET
-  @Path("/appdetail/{app-id}")
+  @Path("/app/{app-id}")
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java:109]: this should also be apps, not app

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java [hunk 27]
```diff
@@ -120,7 +120,7 @@ public void testDeployUsingNonexistantArtifact404() throws Exception {
     Id.Application appId = Id.Application.from(Id.Namespace.DEFAULT, "badapp");
     AppRequest<Config> appRequest =
       new AppRequest<>(new ArtifactSummary("something", "1.0.0"), null);
-    HttpResponse response = deploy(appId, appRequest);
+    HttpResponse response = deploy(appId, appRequest;)
```
Comment (albertshau) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java]: fix

**Review by albertshau — APPROVED** [review]
small comments, otherwise lgtm. Please fix and squash changes before merging
